### PR TITLE
Barcharts!

### PR DESCRIPTION
**Pull request by zwrawr**
image

* Adds barcharts
* Pulls out the legend code from the piechart class so it's more generic and can be reused

### CODE REVIEW

File: src/com/visualization/barchart/bar.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    render( props ) {
+
+        // it's valid for height/width/index to be zero so we have to check it against undefined.
+        if ( !(
+			props &&
+			props['height']	!= undefined &&
+			props['index']	!= undefined &&
+			props['width']	!= undefined &&
+			props['color']
+		) ) {
```
Comment (povrazor) [inline src/com/visualization/barchart/bar.js]: Prefer
```js
if ( !(props && (props.height != null) && (props.index != null) && (props.width != null) && (props.color != null)) ) {
```
or
```js
if ( !(
    props && 
    (props.height != null) &&
    (props.index != null) && 
    (props.width != null) && 
    (props.color != null)
) ) {
```
Noting that:
* The `!=` operator actually checks for both `undefined` and `null`, and `null` is preferred in this case.
* I'm guessing the `props.color` check also needs to be a `!= null` test

Alternatively, if a non-zero/non-empty-string is also a worthwhile thing to be testing for, consider:
```js
if ( !(props && props.height && props.index && props.width && props.color) ) {
```
Though without knowing the code, I could be wrong and `props.index` may be something where `0` should be legal.

File: src/com/visualization/barchart/bar.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            console.warn('Bar was created with invalid props', props);
+            return;
+        }
+
+        let height = props.height;
+        let width = props.width;
+        let index = props.index;
+		let color = props.color;
```
Comment (povrazor) [inline src/com/visualization/barchart/bar.js]: Prefer
```js
let {height, width, index, color} = props;
```

File: src/com/visualization/barchart/barchart.less [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+		align-self: center;
+
+		& > .-svg {
+			& > .-bar {
+				// maybe animate the bar's changing
+			}
+		}
+	}
+}
+
```
Comment (povrazor) [inline src/com/visualization/barchart/barchart.less]: last line should be a newline (not a space then newline)

File: src/com/visualization/legend/legend.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    render( props ) {
+
+        if ( !(
+			props &&
+			props['colors'] &&
+			props['names']
+		) ) {
+            console.warn('Legend was created with invalid props', props);
```
Comment (povrazor) [inline src/com/visualization/legend/legend.js]: prefer
```js
if ( !(props && props.colors && props.names) ) {
```
(notably `props.colors` vs `props['colors']`)

File: src/com/visualization/legend/legend.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            return;
+        }
+
+		let colors = props.colors;
+		let names = props.names;
+
```
Comment (povrazor) [inline src/com/visualization/legend/legend.js]: Prefer
```
let {colors, names} = props;
```

File: src/com/visualization/barchart/bar.js (hunk 1 above)
Comment (zwrawr) [reply src/com/visualization/barchart/bar.js]: 0 is legal for index height and width but not for color.
I keep forgetting that you don't need to use the `props.['key']` syntax

File: src/com/visualization/barchart/barchart.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+export default class BarChart extends Component {
+
+	constructor( props ) {
+		super(props);
+
+    }
+
+
+    scaleValues ( values ) {
```
Comment (povrazor) [inline src/com/visualization/barchart/barchart.js]:
```
scaleValues( values ) {
```
No space between function name and bracket in a declaration

File: src/com/visualization/legend/legend.less [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+	& > .-legend {
+		flex: 1;
+		padding-left: 15px;
+		padding-right: 15px;
+		align-self: center;
+		
+		& li {
+			list-style: none;
+			
+			& .-shape-circle {
```
Comment (povrazor) [inline src/com/visualization/legend/legend.less]:
```
& > .-shape-circle {
```

File: src/com/visualization/legend/legend.less [hunk 8]
```diff
@@ -0,0 +1,29 @@
+
+.chart {
+	& > .-legend {
+		flex: 1;
+		padding-left: 15px;
+		padding-right: 15px;
+		align-self: center;
+		
+		& li {
```
Comment (povrazor) [inline src/com/visualization/legend/legend.less]:
```
& > li {
```

Just a few `>`'s, if we're sure they are always children of the prior element.

File: src/com/visualization/legend/legend.less [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+			
+			& .-shape-circle {
+				display: inline-block;
+				vertical-align: middle;
+				width: 32px;
+				height: 32px;
+				border-radius: 50%;
+			}
+
+			& p {
```
Comment (povrazor) [inline src/com/visualization/legend/legend.less]:
```
& > p {
```

File: src/com/visualization/piechart/piesegment.js [hunk 10]
```diff
@@ -10,7 +10,7 @@ export default class PieSegment extends Component {
     render( props ) {
 
         // it's valid for offset/angle to be zero so we have to check it against undefined.
-        if ( !(props && props['angle'] != undefined && props['offset'] != undefined && props['color']) ) {
+        if ( !(props && props.angle != undefined && props.offset != undefined && props.color) ) {
             console.warn('PieSegment was created with invalid props', props);
```
Comment (povrazor) [inline src/com/visualization/piechart/piesegment.js]: - prefer `!= null` to `!= undefined`.
- wrap any check between `&&` and `||` that deals with multiple items in pair of brackets 
(i.e. `(props.angle != null)`, but not `props` or `props.color`).

**Pull request merged**